runtime/jcl/common/jclvm.c

```diff
@@ -138,6 +150,146 @@ Java_com_ibm_oti_vm_VM_allInstances(JNIEnv * env, jclass unused, jclass clazz, j
+		}
+		hashTableFree(statsTable);
+		numClasses = cursor; /* adjust the length in case the hash table contained nulls */
+		sortStatsTable(statsArray, 0, numClasses - 1);
+	}
+	if (!outOfMemory) {
+		I_32 printedLength = 0;
+		UDATA bufferSize = 0;
+		do {
+			bufferSize += numClasses * 80; /* try incrementally larger sizes */
+			char *stringBuffer = (char *) j9mem_allocate_memory(bufferSize, J9MEM_CATEGORY_VM_JCL);
+			if (NULL == stringBuffer) {
+				outOfMemory = TRUE;
+				break;
+			}
+			printedLength = printHeapStatistics(env, statsArray,numClasses, stringBuffer, bufferSize);
+			if (printedLength > 0) {
+				stringObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(vmThread,
+						(U_8 *) stringBuffer, printedLength, J9_STR_XLAT);
+				stringObjectRef = vmFuncs->j9jni_createLocalRef(env, stringObject);
+			}
+			j9mem_free_memory(stringBuffer);
+		} while (0 == printedLength);
+		j9mem_free_memory(statsArray);
+	}
+
+	if (outOfMemory) {
+		Trc_JCL_heapStatisticsOOM(vmThread);
+		vm->internalVMFunctions->setNativeOutOfMemoryError(vmThread, 0, 0);
+	}
+	vm->internalVMFunctions->internalExitVMToJNI(vmThread);
+
+	return stringObjectRef;
+}
+
+static I_32
+printHeapStatistics(JNIEnv * env,J9HeapStatisticsTableEntry **statsArray,
+		U_32 numClasses, char *stringBuffer, UDATA bufferSize)
+{
+	char *bufferCursor = stringBuffer;
+	I_32 classCursor = 0;
+	UDATA cumulativeCount = 0;
+	UDATA cumulativeSize = 0;
+	UDATA result = 0;
+
+	PORT_ACCESS_FROM_ENV(env);
+
+	result = j9str_printf(PORTLIB, bufferCursor, bufferSize,
+			"%5s%15s%15s    %s\n-------------------------------------------------\n",
+			"num", "object count", "total size", "class name"
+	);
+	bufferCursor += result;
+	bufferSize -= result;
+	for (classCursor = 0; (result > 0) && (classCursor < numClasses); ++classCursor) {
+		J9Class *currentClass = statsArray[classCursor]->clazz;
+		result = j9str_printf(PORTLIB, bufferCursor, bufferSize,
+				"%5d %14lu %14lu    ",
+				classCursor + 1, statsArray[classCursor]->objectCount,
+				statsArray[classCursor]->aggregateSize
+		);
+		bufferCursor += result;
+		bufferSize -= result;
+		if (J9CLASS_IS_ARRAY(currentClass)) {
+			J9ArrayClass *arrayClazz = (J9ArrayClass*)currentClass;
+			UDATA arity = arrayClazz->arity;
+			J9Class *leafComponentType = arrayClazz->leafComponentType;
+			J9ROMClass * leafROMClass = leafComponentType->romClass;
+			J9UTF8 *leafName = J9ROMCLASS_CLASSNAME(leafROMClass);
+			UDATA isPrimitive = J9ROMCLASS_IS_PRIMITIVE_TYPE(leafROMClass);
+			UDATA i = 0;
+			for (i = 0; i < arity; ++i) {
+				result = j9str_printf(PORTLIB, bufferCursor, bufferSize, "[");
+				bufferCursor += result;
+				bufferSize -= result;
+			}
+			if (isPrimitive) {
+				result = j9str_printf(PORTLIB, bufferCursor, bufferSize, "%c\n",
+						J9UTF8_DATA(J9ROMCLASS_CLASSNAME(leafComponentType->arrayClass->romClass))[1]);
+			} else {
+				result = j9str_printf(PORTLIB, bufferCursor, bufferSize, "L%.*s;\n",
```

Comment:
Why `"L%.*s;\n"` here, but `"%.*s\n"` below?

Reply:
That is for compatibility with the reference implementation: isolated objects are shown as the plain fully qualified name, while arrays add use the `L...;` format:

>   93:             1             80  [Ljava.lang.ThreadLocal$ThreadLocalMap$Entry;
>   94:             2             80  [Ljava.net.URL;
>   95:             2             80  java.io.BufferedWriter
>   96:             2             80  java.io.ExpiringCache
>